src/three-components/PMREMGenerator.ts

```diff
@@ -0,0 +1,367 @@
+      // This hack comes from the original PMREMGenerator; if you set envMap
+      // (even though it doesn't exist on ShaderMaterial), the assembled
+      // shader will populate the correct function into envMapTexelToLinear().
+      (material as any).envMap = target.texture;
+      material.uniforms['faceIndex'].value = k;
+
+      let planeMesh = new Mesh(plane, material);
+
+      planeMesh.position.x = (0.5 + (k % 3)) * sizePad;
+      planeMesh.position.y = (0.5 + (k > 2 ? 1 : 0)) * sizePad + offset;
+      (planeMesh.material as Material).side = DoubleSide;
+      planeMesh.scale.setScalar(sizePad);
+      this.objects.push(planeMesh);
+    }
+  }
+
+  updateLodObjects(index: number, target: WebGLRenderTargetCube) {
+    for (let k = index * 6; k < (index + 1) * 6; k++) {
+      const material = (this.objects[k].material as any);
+      material.uniforms['envMap'].value = target.texture;
+      material.envMap = target.texture;
+    }
+  }
+
+  generateMipmaps(cubeTarget: WebGLRenderTargetCube) {
+    const gammaInput = this.renderer.gammaInput;
+    const gammaOutput = this.renderer.gammaOutput;
+    const toneMapping = this.renderer.toneMapping;
+    const toneMappingExposure = this.renderer.toneMappingExposure;
+    const currentRenderTarget = this.renderer.getRenderTarget();
+
+    this.renderer.toneMapping = LinearToneMapping;
+    this.renderer.toneMappingExposure = 1.0;
+    this.renderer.gammaInput = false;
+    this.renderer.gammaOutput = false;
+
+    this.mipmapShader.uniforms['texelSize'].value = 1.0 / cubeTarget.width;
+    this.mipmapShader.uniforms['envMap'].value = cubeTarget.texture;
+    (this.mipmapShader as any).envMap = cubeTarget.texture;
+    for (let i = this.numLods - 1; i >= 0; i--) {
+      this.cubeCamera.renderTarget = this.cubeLods[i];
+      this.cubeCamera.update(this.renderer, this.mipmapScene);
+      this.mipmapShader.uniforms['texelSize'].value =
+          1.0 / this.cubeLods[i].width;
+      this.mipmapShader.uniforms['envMap'].value = this.cubeLods[i].texture;
+      (this.mipmapShader as any).envMap = this.cubeLods[i].texture;
+    }
+
+    this.renderer.setRenderTarget(currentRenderTarget);
+    this.renderer.toneMapping = toneMapping;
+    this.renderer.toneMappingExposure = toneMappingExposure;
+    this.renderer.gammaInput = gammaInput;
+    this.renderer.gammaOutput = gammaOutput;
+  }
+
+  packMipmaps() {
+    this.objects.forEach((object) => {
+      this.packingScene.add(object);
+    });
+
+    const gammaInput = this.renderer.gammaInput;
+    const gammaOutput = this.renderer.gammaOutput;
+    const toneMapping = this.renderer.toneMapping;
+    const toneMappingExposure = this.renderer.toneMappingExposure;
+    const currentRenderTarget = this.renderer.getRenderTarget();
+
+    this.renderer.gammaInput = false;
+    this.renderer.gammaOutput = false;
+    this.renderer.toneMapping = LinearToneMapping;
+    this.renderer.toneMappingExposure = 1.0;
```

Comment:
I wonder if we could dry this class up a bit by adding a pair of "save" / "restore" helpers for un-setting and re-setting the state of the renderer..